dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/bundle/DefaultTrackerBundleService.java

```diff
@@ -228,6 +233,69 @@ public TrackerBundleReport commit( TrackerBundle bundle )
         return bundleReport;
     }
 
+    @Override
+    @Transactional
+    public TrackerBundleReport delete( TrackerBundle bundle )
+    {
+        TrackerBundleReport bundleReport = new TrackerBundleReport();
+
+        if ( TrackerBundleMode.VALIDATE == bundle.getImportMode() )
+        {
+            return bundleReport;
+        }
+
+        TrackerTypeReport trackedEntityReport = deleteTrackedEntities( bundle );
+        TrackerTypeReport enrollmentReport = deleteEnrollments( bundle );
+        TrackerTypeReport eventReport = deleteEvents( bundle );
+        TrackerTypeReport relationshipReport = deleteRelationShips( bundle );
+
+        bundleReport.getTypeReportMap().put( TrackerType.TRACKED_ENTITY, trackedEntityReport );
+        bundleReport.getTypeReportMap().put( TrackerType.ENROLLMENT, enrollmentReport );
+        bundleReport.getTypeReportMap().put( TrackerType.EVENT, eventReport );
+        bundleReport.getTypeReportMap().put( TrackerType.RELATIONSHIP, relationshipReport );
+
+        dbmsManager.clearSession();
+        cacheManager.clearCache();
+
+        return bundleReport;
+    }
+
```

Comment:
Suggestion:
to remove the duplication of code here you can create a map to link the TrackerType to the function to delete, then use iterate over this map to run the delete method, that will take the TrackedType, the function and the bundle as parameters. This way, we will have just one delete method instead of 4.